03-first-task.Rmd

```diff
@@ -3,42 +3,418 @@ ottrpal::set_knitr_image_path()
+
+Thankfully, the solution here is simple: Move all of the input files directly into the working directory.
+
+```
+task BwaMem {
+  input {
+    File input_fastq
+    File ref_fasta
+    File ref_fasta_index
+    File ref_dict
+    File ref_amb
+    File ref_ann
+    File ref_bwt
+    File ref_pac
+    File ref_sa
+    Int threads = 16
+  }
+
+  command <<<
+    set -eo pipefail
+
+    # This can also be done by creating an array and then looping that array,
+    # but we'll do it one line at a time or clarity's sake.
+    mv "~{ref_fasta}" .
+    mv "~{ref_fasta_index}" .
+    mv "~{ref_dict}" .
+    mv "~{ref_amb}" .
+    mv "~{ref_ann}" .
+    mv "~{ref_bwt}" .
+    mv "~{ref_pac}" .
+    mv "~{ref_sa}" .
+
+    bwa mem \
+    [...]
+  >>>
+}
+```
+::: {.notice data-latex="warning"}
+Some backends/executors do not support `mv` acting on input files. If you are running into problems with this and are working with miniwdl, the `--copy-input-files` flag will usually allow `mv` to work. You could also simply use `cp` to copy the files instead of move them, although this may not be an efficient use of disk space, so consider using `mv` if your target backends and executors can handle it.
+:::
+
+With our files now all in the working directory, we can turn our attention to the bwa task itself. We can no longer directly pass in `~{ref_fasta}` or `~{input_fastq}`, because those variables will point to a non-existent file in a now-empty input directory. For the reference fasta, if we know for sure that the user will always pass in a file named `ref.fa`, we could hardcode that. For the input fastq, we can use the built-in bash function basename. That would leave us with something like this:
+
+```
+# create a bash variable named fastq which is the basename of the input fastq
+# be aware that this variable only exists in the scope of this task's command section
+fastq="$(basename -- ~{input_fastq})"
+
+# actually run bwa mem
+bwa mem \
+      -p -v 3 -t ~{threads} -M -R '@RG\tID:foo\tSM:foo2' \
+      ref.fa $fastq > my_nifty_output.sam 
+```
+
+However, we may want to avoid assuming that the reference genome is always called ref.fa. We may also want to avoid using a mixture of bash variables and variables-passed-in-from WDL, as keep track of the variables may become difficult. Thankfully, we can use private input variables to address this scenario.
```

Comment:
I'm thinking of leaving this out, because you provide a much cleaner solution below via private variables.

Reply:
I'm confused why we can't just use `~{input_fastq}` here.

Reply:
> I'm confused why we can't just use ~{input_fastq} here.

In an earlier version I also mv'd the input fastq because I forgot that the redirect would put it in the workdir. I'll remove that reference, thanks for catching this!


> I'm thinking of leaving this out, because you provide a much cleaner solution below via private variables.

How about this new version:

---

With our files now all in the working directory, we can turn our attention to the bwa task itself. We can no longer directly pass in `~{ref_fasta}` or any of the other files we `mv`ed into the working directory, because those variables will point to a non-existent file in a now-empty input directory. There are several ways to solve this problem: 
* Assuming the filename of an input is constant, which might be a safe assumption for reference files
* Using the bash built-in `basename` function
* Using the WDL built-in `basename()` function along with private variables

We recommend using the last option, as it works for essentially any input and may be more intuitive than the bash `basename` function. The next section will provide an example.

Reply:
Nice.